### ISSUE TITLE
All-NaN gradients from preconditioned `jax.scipy.sparse.linalg.cg` for small cotangent norms

### ISSUE THREAD

**Issue by dionhaefner** [jax-ml/jax#40254]
## Summary

`jax.grad` through `jax.scipy.sparse.linalg.cg` with a preconditioner can return an all-NaN gradient while the primal solve converges cleanly, with nothing signaling the failure.

The cause is a scale mismatch that is specific to reverse-mode AD. `_isolve` reuses the forward solver as the transpose solve:

```python
x = lax.custom_linear_solve(A, b, solve=isolve_solve, transpose_solve=isolve_solve,
                            symmetric=symmetric)
```

so the backward pass solves the same preconditioned system, but its right-hand side is the incoming *cotangent*. The cotangent's norm is set by whatever the surrounding program computes downstream of the solve, and for a scalar loss over a large field it is routinely many orders of magnitude below the primal rhs (1e-13 in our pipeline, with a primal rhs of order 100). At that scale the float32 inner products inside `_cg_solve` underflow and the loop divides 0/0.

This is not easily fixable (or even debuggable) by the user; the only workaround we found is to rebuild the `custom_linear_solve` wrapper ourselves so we can normalize inside it (shown below).

## Reproducer

CPU, float32, fails jitted and eager. The primal system is well scaled. The loss carries a small scale (1e-13, standing in for whatever units and composition induce downstream), so the cotangent reaching the transpose solve has norm ~1e-13. 3 of the first 10 seeds fail this way on my machine, seed 5 shown.

```python
import numpy as np
import jax
import jax.numpy as jnp

N, IH2 = 32, 3.0 * 32**2

# smooth random level-set -> two-phase permeability with 1e6 contrast
rng = np.random.default_rng(5)
x = np.linspace(0, 2 * np.pi, N, endpoint=False)
X, Y, Z = np.meshgrid(x, x, x, indexing="ij")
f = np.zeros((N, N, N))
for _ in range(6):
    k = rng.integers(1, 4, size=3)
    f += rng.normal() * np.cos(k[0] * X + rng.uniform(0, 2 * np.pi)) \
         * np.cos(k[1] * Y + rng.uniform(0, 2 * np.pi)) \
         * np.cos(k[2] * Z + rng.uniform(0, 2 * np.pi))
level = jnp.asarray((f / np.abs(f).max()).astype(np.float32))
free = np.ones((N, N, N), dtype=bool)
free[:, :, 0] = free[:, :, -1] = False   # fixed z-faces
free = jnp.asarray(free)
b = jnp.where(free, jnp.asarray(rng.normal(size=(N, N, N)).astype(np.float32)), 0.0)

LOSS_SCALE = 1e-13  # stand-in for the scale the downstream computation induces


def loss(level, use_precond):
    kap = 1e-6 + jax.nn.sigmoid(-8.0 * level)

    def op(p):  # variable-coefficient diffusion, harmonic-mean face coefficients
        p = jnp.where(free, p, 0.0)
        out = jnp.zeros_like(p)
        for ax in range(3):
            for sh in (-1, 1):
                knb, pnb = jnp.roll(kap, sh, ax), jnp.roll(p, sh, ax)
                edge = tuple((-1 if sh == -1 else 0) if a == ax else slice(None) for a in range(3))
                valid = jnp.ones_like(p, dtype=bool).at[edge].set(False)
                fk = 2.0 * kap * knb / (kap + knb + 1e-12)
                out = out + jnp.where(valid, fk * IH2, 0.0) * (p - pnb)
        return jnp.where(free, out, p)

    if use_precond:  # Jacobi
        diag = jnp.zeros_like(kap)
        for ax in range(3):
            for sh in (-1, 1):
                knb = jnp.roll(kap, sh, ax)
                edge = tuple((-1 if sh == -1 else 0) if a == ax else slice(None) for a in range(3))
                valid = jnp.ones_like(kap, dtype=bool).at[edge].set(False)
                diag = diag + jnp.where(valid, 2.0 * kap * knb / (kap + knb + 1e-12) * IH2, 0.0)
        inv = 1.0 / jnp.where(free, diag, 1.0)
        M = lambda r: jnp.where(free, inv * r, r)
    else:
        M = None

    sol, _ = jax.scipy.sparse.linalg.cg(op, b, tol=1e-7, maxiter=1000, M=M)
    return LOSS_SCALE * jnp.sum(jnp.where(free, sol, 0.0))


for use_precond in (True, False):
    val, grad = jax.jit(jax.value_and_grad(loss), static_argnums=1)(level, use_precond)
    print(f"M={'Jacobi' if use_precond else 'None  '}:  primal loss = {float(val):.6e} (finite)   "
          f"grad finite = {bool(jnp.isfinite(grad).all())}")
```

Output (CPU, jax 0.10.2):

```
M=Jacobi:  primal loss = 1.300677e-12 (finite)   grad finite = False
M=None  :  primal loss = 1.300678e-12 (finite)   grad finite = True
```

## Mechanism

In `_cg_solve`, when `M` is given the loop condition tests the unpreconditioned residual, but the body divides by the preconditioned inner products:

```python
def cond_fun(value):
    _, r, gamma, _, k = value
    rs = gamma.real if M is _identity else _vdot_real_tree(r, r)
    return (rs > atol2) & (k < maxiter)

def body_fun(value):
    ...
    alpha = gamma / _vdot_real_tree(p, Ap).astype(dtype)
    ...
    beta_ = gamma_ / gamma
```

On a cotangent-scale system, `gamma = r.z` is a sum of nonnegative products (`inv * r**2`) whose terms sit near 1e-40 as the residual shrinks. When the sum reaches exact zero while `r.r` is still above `atol2`, the loop keeps running and computes `beta_ = 0/0`. Without `M`, `cond_fun` tests `gamma` itself and exits before the division, so the unpreconditioned solve degrades gracefully (early exit) instead of NaNing on the same rhs.

## A possible workaround

Wrapping the entire solver in `custom_linear_solve` and re-scaling the rhs only on the backwards pass fixes the issue:

```python
def cg_scale_invariant(A, b, *, tol, maxiter, M):
    def solve(matvec, rhs):
        nrm = jnp.sqrt(jnp.sum(rhs * rhs))
        scale = jnp.where(nrm > 0, 1.0 / nrm, 1.0)
        x, _ = jax.scipy.sparse.linalg.cg(matvec, rhs * scale, tol=tol, maxiter=maxiter, M=M)
        return x / scale
    return lax.custom_linear_solve(A, b, solve=solve, transpose_solve=solve, symmetric=True)
```

Substituting this for the `cg` call in the reproducer gives a bit-identical primal loss and a finite gradient.

## Proposed fixes

**1. Rescale the transpose solve.** This is the workaround above, applied inside `_isolve` where the transpose solve actually lives. CG is scale-invariant in exact arithmetic and the tolerance is relative, so solving the transpose system at unit rhs norm and rescaling the result changes nothing semantically, but it keeps the backward solve in the range float32 can handle regardless of the cotangent's scale. In `_isolve`:

```python
isolve_solve = partial(_isolve_solve, x0=x0, tol=tol, atol=atol, maxiter=maxiter, M=M)

def transpose_solve_scaled(vecmat, b):
    nrm = jnp.sqrt(_vdot_real_tree(b, b))
    scale = jnp.where(nrm > 0, 1 / nrm, 1.0)
    x = _isolve_solve(vecmat, _mul(scale, b), x0=_mul(scale, x0), tol=tol,
                      atol=atol * scale, maxiter=maxiter, M=M)
    return _mul(1 / scale, x)

x = lax.custom_linear_solve(A, b, solve=isolve_solve,
                            transpose_solve=transpose_solve_scaled, symmetric=symmetric)
```

**2. Add a fail-safe: test `gamma` in `cond_fun` unconditionally.**

```python
return (rs > atol2) & (k < maxiter) & (gamma.real > 0)
```

This only affects executions that currently return NaN: with an SPD preconditioner `gamma` is a sum of nonnegative products, and once it is exactly zero, `beta_ = gamma_ / 0` poisons `p` and the iterate goes NaN within two iterations. (For unpreconditioned solves this already exits at `gamma == 0`, since there `rs` *is* `gamma`.) Exiting instead returns the best iterate float32 could reach. This guarantees a finite result but not a converged one, so it complements rather than replaces fix 1.

**3. Document the issue, and recommend to account for cotangent scale.** E.g. a note on `cg`/`bicgstab` that the built-in VJP re-runs the solver on the cotangent as rhs, that float32 preconditioned solves can break down at small cotangent scales, and the wrapper from the workaround section as a potential escape hatch.

Happy to contribute a PR if any of those options sound appealing.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.10.2
jaxlib: 0.10.2
numpy:  2.4.6
python: 3.14.6 (main, Jun 10 2026, 10:03:53) [Clang 21.0.0 (clang-2100.0.123.102)]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Darwin', node='MacBookPro', release='25.5.0', version='Darwin Kernel Version 25.5.0: Tue Jun  9 22:18:58 PDT 2026; root:xnu-12377.121.10~1/RELEASE_ARM64_T6000', machine='arm64')
```
